File: src/contests/cg/puzzle/pascal/mirrors.pas
// 🎮 CodinGame Puzzle - mirrors
// https://www.codingame.com/training/easy/mirrors

{$mode objfpc}{$H+}
var
  n, i: integer;
  r: array of double;
  reflected, ri, denom: double;
begin
  readln(n);
  setlength(r, n);
  for i := 0 to n - 1 do read(r[i]);
  readln;
  reflected := 0.0;
  for i := n - 1 downto 0 do
  begin
    ri := r[i];
    denom := 1 - ri * reflected;
    if denom = 0 then reflected := ri
    else reflected := ri + ((1 - ri) * (1 - ri) * reflected) / denom;
  end;
  writeln(reflected:0:4);
end.

File: src/contests/cg/puzzle/ocaml/mirrors.ml
(* 🎮 CodinGame Puzzle - mirrors *)
(* https://www.codingame.com/training/easy/mirrors *)

let () =
  let _ = read_line () in
  let parts = String.split_on_char ' ' (String.trim (read_line ())) in
  let r = Array.of_list (List.map float_of_string parts) in
  let reflected = ref 0.0 in
  for i = Array.length r - 1 downto 0 do
    let ri = r.(i) in
    let denom = 1.0 -. ri *. !reflected in
    reflected :=
      ri +. (if denom = 0.0 then 0.0
             else ((1.0 -. ri) *. (1.0 -. ri) *. !reflected) /. denom)
  done;
  Printf.printf "%.4f\n" !reflected
